### ISSUE TITLE
Access violation in `impl.loadExtraGL()`

### ISSUE THREAD

**Issue by Straivers** [DerelictOrg/DerelictGL3#56]
I am following the instructions in the docs for creating a custom context, but have encountered an issue where an access violation occurs during `DerelictGL3.load()`. If I do not use the custom context mechanism, `DerelictGL3.load()` occurs without issue.

I'm not certain that it's not my fault, but any insight would be appreciated.

My current context code is:

```D
//This does not support depreciated imports
mixin(glImports);

//Import type declarations for the highest supported version of
//OpenGL, excluding depreciations
mixin glDecls!(GLVersion.highestSupported);

struct GLContext(GLVersion contextVersion)
{
    //Core OpenGL functions
    mixin glFuncs!contextVersion;
    mixin glLoaders!contextVersion;
}
```

[The crash message is attached](https://github.com/DerelictOrg/DerelictGL3/files/1218978/derelictgl3_access_violation_loadExtraGL.txt)

**Comment by mdparker** [comment]
Did you instantiate an instance of your GLContext type? You'll need at least one instance of it.

```
auto myContext = GLContext!(GLVersion.highestSupported);
```

**Comment by Straivers** [comment]
Yes, it was.

**Comment by mdparker** [comment]
I can't reproduce this on Windows, compiling the following with `dub -ax86_64 --single app.d`, with the 64-bit glfw3.dll in the same directory. Can you try this out? And also, if you can't find any problems with your configuration, provide a minimal version that reproduces the problem? The error output alone is not enough for me to do anything more than make blind guesses since I can't reproduce it.

```d
#!user/bin/env dub
/+ dub.sdl:
name "hellowindow"
dependency "derelict-glfw3" version="~>4.0.0-beta"
dependency "derelict-gl3" path="../../gl3"
versions "DerelictGL3_Contexts"
+/

import std.stdio;
import derelict.glfw3;
import derelict.opengl;

mixin(glImports);
mixin glDecls!(GLVersion.highestSupported);

struct GLContext(GLVersion contextVersion) {
    mixin glFuncs!contextVersion;
    mixin glLoaders!contextVersion;
}

GLContext!(GLVersion.highestSupported) ctx;

enum uint _winWidth = 1024;
enum uint _winHeight = 768;

void main()
{
    writeln("Loading libraries...");
    DerelictGL3.load();
    DerelictGLFW3.load();

    writeln("Initializing GLFW...");
    if(!glfwInit()) throw new Exception("Failed to initialize GLFW");
    scope(exit) glfwTerminate();

    writeln("Creating window...");
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 5);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
    glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
    glfwWindowHint(GLFW_RESIZABLE, GL_FALSE);

    auto window = glfwCreateWindow(_winWidth, _winHeight, "LearnOpenGLinD", null, null);
    if(!window) throw new Exception("Failed to create GLFW window");

    glfwMakeContextCurrent(window);
    glfwSetKeyCallback(window, &keyCallback);

    ctx.load();

    int w, h;
    glfwGetFramebufferSize(window, &w, &h);
    ctx.glViewport(0, 0, w, h);

    ctx.glClearColor(0.2f, 0.3f, 0.3f, 1.0f);

    writeln("Entering main loop...");
    while(!glfwWindowShouldClose(window)) {
        glfwPollEvents();
        ctx.glClear(GL_COLOR_BUFFER_BIT);

        glfwSwapBuffers(window);
    }
}

extern(C) nothrow
void keyCallback(GLFWwindow* window, int key, int scanCode, int action, int mode)
{
    if(key == GLFW_KEY_ESCAPE && action == GLFW_PRESS) {
        glfwSetWindowShouldClose(window, GL_TRUE);
    }
}```

**Comment by Straivers** [comment]
I notice that you call `DerelictGL3.load()` outside of the context struct? Does that have an impact?

This is what I've got:

```d
/+ dub.sdl:
name "hello"
dependency "derelict-gl3" version="~>2.0.0-beta"
dependency "oswald" version="0.3.0"
versions "DerelictGL3_Contexts"
+/

module test;

pragma(lib, "opengl32");
pragma(lib, "gdi32");

import core.sys.windows.windows;
import derelict.opengl;
import oswald;

import std.typecons: Yes;

/* Both cause the same error */
// enum ctxVer = GLVersion.highestSupported;
enum ctxVer = GLVersion.gl11;

mixin(glImports);
mixin glDecls!(ctxVer);

struct Context
{
    mixin glFuncs!(ctxVer);
    mixin glLoaders!(ctxVer);

    HGLRC ctx;
    HDC dc;

    this(HWND hwnd)
    {
        dc = GetDC(hwnd);
        ctx = createContext(hwnd, dc);
        DerelictGL3.load();
    }
}

void main()
{
    OsWindow window;
    OsWindow.createNew(WindowConfig("hello", 720, 480), &window);

    window.input.keyCallback = (window, key) {
        if (key.keycode == Keycodes.Escape)
            window.close();
    };

    auto context = Context(window.platformData.handle);

    wglMakeCurrent(context.dc, context.ctx);
    context.load();
    
    window.input.process(Yes.waitForEvents);
}

HGLRC createContext(HWND hwnd, HDC hdc)
{
    PIXELFORMATDESCRIPTOR pfd;
    pfd.nSize = PIXELFORMATDESCRIPTOR.sizeof;
    pfd.nVersion = 1;
    pfd.dwFlags = PFD_DRAW_TO_WINDOW | PFD_SUPPORT_OPENGL | PFD_DOUBLEBUFFER;
    pfd.iPixelType = PFD_TYPE_RGBA;
    pfd.cColorBits = 24;

    const pixelFormat = ChoosePixelFormat(hdc, &pfd);
    if (SetPixelFormat(hdc, pixelFormat, &pfd) == FALSE)
        assert(false, "Failed to set pixel format for temp context");

    HGLRC context = wglCreateContext(hdc);
    if (context is null)
    {
        import std.conv : to;

        assert(false, "Failed to create temp context: Error Code " ~ GetLastError()
                .to!string);
    }

    if (wglMakeCurrent(hdc, context) == FALSE)
    {
        wglDeleteContext(context);
        assert(false, "Failed to make temp context current");
    }

    return context;
}
```

**Comment by mdparker** [comment]
Yes. It's what loads the OpenGL shared library into memory -- not the functions, it just gets a handle to the library (via `LoadLibrary` on Windows and `dlopen` elsewhere) so the functions can then be loaded. Without that, none of the OpenGL functions can be loaded. You need the context to know which version of OpenGL to load, but you don't need the context to get a handle to the library.

**Comment by mdparker** [comment]
OK. I didn't look at your code before I answered your question. Having now done so, I see what's going on. You're linking to OpenGL, which is fine. That allows you to use the `wgl*` stuff directly. However, it does nothing for the derelict library because it expects the library to be loaded internally dynamically, no matter what happens in your own code. You still need to call `DerelictGL3.load()` so it can line up the internal function pointers it uses with the symbols in the library.

**Comment by mdparker** [comment]
Sorry, ignore my previous comments. I missed your call to `load` in the context constructor. I need to run this on my Windows box to see what's going on. I'm on my MacBook at the moment.

**Comment by mdparker** [comment]
Here's what I know so far. On Windows dmd (haven't tested elsewhere yet), it only happens when compiling with the default 32-bit (-m32), and only when specifying `DerelictGL3_Contexts`. It happens with your code, my code, and when using the default `glContext` mixin template. Given all the mixed in code, I think it's going to take me a bit of time to work this out. In the meantime, if you can compile as 64-bit, that may be a workaround for you.

**Comment by mdparker** [comment]
An issue reported in the forums appears to be related: http://forum.dlang.org/thread/mnjgsdyvcrvwpaiyryvz@forum.dlang.org

**Comment by igor84** [comment]
I may have found the reason. The file glloader.d around line 300 has this:
```
    static if(!Derelict_OS_Mac) {
        extern(C) @nogc nothrow alias da_getProcAddress = void* function(const(char)*);
        __gshared da_getProcAddress getProcAddress;
    }
```
When I just changed `extern(C)` to `extern(Windows)` I stopped getting the crashes on DMD Release and LDC 32 bit builds.

**Comment by mdparker** [comment]
Good catch! Thanks for the help. I'll get a fix pushed in a few hours.

**Comment by mdparker** [comment]
Apologies. A few hours became a few days, but it's fixed in 2.0.0-beta.4. Thanks again!
